### PR TITLE
Extend Functions to printout code

### PR DESCRIPTION
**Pull request by IvanKolesarEquinor**

### CODE REVIEW

**Comment by JensGM** [comment]
Sorry, took a while to get to this, this is really nice, but I'm wondering if we'd manage to make it even nicer with some extra _features_:

* Could we make it so that the code object not only supplies sources, but maybe handles compilation and assembly too?
* The tests are really nice, but I can imagine them being a hassle to maintain in the future, maybe it's an idea to parse the source code generated from the properties? And compare ASTs, rather than text?

File: xun/tests/test_functions.py [hunk 1]
```diff
@@ -457,6 +457,63 @@ def h():
             return f
 
 
+def test_functions_code():
```
Comment (JensGM) [inline xun/tests/test_functions.py:460]: Remove test, we don't need it

File: xun/functions/function.py [hunk 2]
```diff
@@ -58,11 +59,44 @@ def __init__(self, desc, dependencies, max_parallel):
...
         self._graph_builder = None
+        self._code = None
 
     @property
     def name(self):
         return self.desc.name
 
+    @property
+    def code(self):
+        class FunctionCode:
```
Comment (JensGM) [inline xun/functions/function.py:70]: Make this class a nested class inside function

File: xun/functions/function.py [hunk 3]
```diff
@@ -58,11 +59,44 @@ def __init__(self, desc, dependencies, max_parallel):
...
         self.hash = Function.sha256(desc, dependencies)
         self._graph_builder = None
+        self._code = None
 
     @property
     def name(self):
         return self.desc.name
 
+    @property
+    def code(self):
```
Comment (JensGM) [inline xun/functions/function.py:69]: Keep only one instance, not property

**Review by JensGM — CHANGES_REQUESTED** [review]

File: xun/functions/function.py [hunk 4]
```diff
@@ -58,11 +59,44 @@ def __init__(self, desc, dependencies, max_parallel):
...
+        class FunctionCode:
+            def __init__(self, func):
+                self.__func = func
```
Comment (JensGM) [inline xun/functions/function.py:72]:
```suggestion
                self.owner = func
```

File: xun/tests/test_functions.py (hunk 1 above)
Comment (IvanKolesarEquinor) [reply xun/tests/test_functions.py]: Done

File: xun/functions/function.py (hunk 2 above)
Comment (IvanKolesarEquinor) [reply xun/functions/function.py]: Done

File: xun/functions/function.py (hunk 3 above)
Comment (IvanKolesarEquinor) [reply xun/functions/function.py]: Done

File: xun/functions/function.py (hunk 4 above)
Comment (IvanKolesarEquinor) [reply xun/functions/function.py]: owner it is

**Review by JensGM — APPROVED** [review]
Looks good to me!

**Comment by JensGM** [comment]
Rebase and merge at will

**Pull request merged**